dds/DCPS/XTypes/DynamicDataImpl.cpp

```diff
@@ -5962,6 +5964,637 @@ bool operator<<(Serializer& ser, const KeyOnly<DDS::DynamicData_ptr>& key)
   return serialize(ser, key.value, Sample::KeyOnly);
 }
 
+// vwrite function for dynamic data.
+
+bool check_rc(DDS::ReturnCode_t rc, DDS::MemberId id, DDS::TypeKind tk, const char* fn_name)
+{
+  return check_rc_from_get(rc, id, tk, fn_name, LogLevel::Warning);
+}
+
+void vwrite(ValueWriter& vw, DDS::DynamicData_ptr value);
+
+void vwrite_item(ValueWriter& vw, DDS::DynamicData_ptr value,
+                 DDS::MemberId id, const DDS::DynamicType_var& item_type)
+{
+  using namespace XTypes;
+  const DDS::TypeKind item_tk = item_type->get_kind();
+  DDS::TypeKind tk = item_tk;
+
+  DDS::ReturnCode_t rc = DDS::RETCODE_OK;
+  if (item_tk == TK_ENUM) {
+    rc = enum_bound(item_type, tk);
+    if (rc != DDS::RETCODE_OK) {
+      if (log_level >= LogLevel::Warning) {
+        ACE_ERROR((LM_WARNING, "(%P|%t) WARNING: vwrite_item: enum_bound failed (%C)\n",
+                   retcode_to_string(rc)));
+      }
+      return;
+    }
+  } else if (item_tk == TK_BITMASK) {
+    rc = bitmask_bound(item_type, tk);
+    if (rc != DDS::RETCODE_OK) {
+      if (log_level >= LogLevel::Warning) {
+        ACE_ERROR((LM_WARNING, "(%P|%t) WARNING: vwrite_item: bitmask_bound failed (%C)\n",
+                   retcode_to_string(rc)));
+      }
+      return;
+    }
+  }
+
+  switch (tk) {
+  case TK_INT8: {
+    CORBA::Int8 val = 0;
+    rc = value->get_int8_value(val, id);
+    if (!check_rc(rc, id, tk, "vwrite_item")) {
+      return;
+    }
+    if (rc == DDS::RETCODE_OK) {
+      if (item_tk == TK_ENUM) {
+        write_enum(vw, item_type, val);
+      } else {
+        vw.write_int8(val);
+      }
+    } else {
```

Comment:
Should these explicitly check for NO_DATA?

Reply:
At this point, we're sure that it's either OK or NO_DATA, so it doesn't need to check.